### PR TITLE
Replace NumPy with Torch in examples/fabric/

### PR DESCRIPTION
**Pull request by ryan597**
## What does this PR do?

Refactoring to remove all numpy function calls in the `examples/fabric` folder.

Removal of numpy package import in the listed files.
Most replacements are simple substitutions of `np.<func>` to `torch.<func>`, but also the inclusion of logical_not() where subtraction with tensors of bools are not allowed.
Removal of seeding numpy rng
Where `.numpy()` is called, it has been changed to `.tolist()` as input to external functions.

Files:
- `examples/fabric/meta_learning/train_fabric.py`
- `examples/fabric/meta_learning/train_torch.py`
- `examples/fabric/reinforcement_learning/rl/agent.py`
- `examples/fabric/reinforcement_learning/rl/utils.py`
- `examples/fabric/reinforcement_learning/train_torch.py`
- `examples/fabric/reinforcement_learning/train_fabric.py`
- `examples/fabric/reinforcement_learning/train_fabric_decoupled.py`

Part of #16649 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by ryan597** [comment]
lightning-fabric (GPUs) (testing pkg: fabric) is failing at the standalone test for `strategies/test_fsdp_integration.py`.
[Link to pipeline fail](https://dev.azure.com/Lightning-AI/lightning/_build/results?buildId=148381&view=logs&j=3f274fac-2e11-54ca-487e-194c91f3ae9f&t=8da0b6c4-35de-5c68-3c64-0e4ce0eb560c&l=637) 

I'm not sure how to go about resolving the failed test as the examples I changed aren't going into that test..

File: examples/fabric/reinforcement_learning/train_torch.py [hunk 1]
```diff
@@ -204,9 +202,9 @@ def main(args: argparse.Namespace):
 
             # Single environment step
             next_obs, reward, done, truncated, info = envs.step(action.cpu().numpy())
-            done = np.logical_or(done, truncated)
+            done = torch.logical_or(Tensor(done), Tensor(truncated))
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/train_torch.py:205]:
```suggestion
            done = torch.logical_or(torch.tensor(done), torch.tensor(truncated))
```

File: examples/fabric/reinforcement_learning/train_torch.py [hunk 2]
```diff
@@ -204,9 +202,9 @@ def main(args: argparse.Namespace):
...
             rewards[step] = torch.tensor(reward, device=device).view(-1)
-            next_obs, next_done = Tensor(next_obs).to(device), Tensor(done).to(device)
+            next_obs, next_done = Tensor(next_obs).to(device), done.to(device)
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/train_torch.py:207]:
```suggestion
            next_obs, next_done = torch.tensor(next_obs, device=device), done.to(device)
```

File: examples/fabric/reinforcement_learning/train_fabric_decoupled.py [hunk 3]
```diff
@@ -124,9 +123,9 @@ def player(args, world_collective: TorchCollective, player_trainer_collective: T
 
             # Single environment step
             next_obs, reward, done, truncated, info = envs.step(action.cpu().numpy())
-            done = np.logical_or(done, truncated)
+            done = torch.logical_or(Tensor(done), Tensor(truncated))
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/train_fabric_decoupled.py:126]: same here :)

File: examples/fabric/reinforcement_learning/train_fabric.py [hunk 4]
```diff
@@ -150,9 +149,9 @@ def main(args: argparse.Namespace):
 
             # Single environment step
             next_obs, reward, done, truncated, info = envs.step(action.cpu().numpy())
-            done = np.logical_or(done, truncated)
+            done = torch.logical_or(Tensor(done), Tensor(truncated))
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/train_fabric.py:152]: same :)

File: examples/fabric/reinforcement_learning/rl/agent.py [hunk 5]
```diff
@@ -24,7 +23,8 @@ def __init__(self, envs: gym.vector.SyncVectorEnv, act_fun: str = "relu", ortho_
             raise ValueError("Unrecognized activation function: `act_fun` must be either `relu` or `tanh`")
         self.critic = torch.nn.Sequential(
             layer_init(
-                torch.nn.Linear(np.array(envs.single_observation_space.shape).prod(), 64), ortho_init=ortho_init
+                torch.nn.Linear(int(torch.Tensor(envs.single_observation_space.shape).prod()), 64),
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/rl/agent.py:26]: same here, we should always use `torch.tensor()` in favor of torch.Tensor(). I know it's weird but it's what torch recommends.

**Review by awaelchli — COMMENTED** [review]
LGTM :) 
minor comments regarding the creation of tensors.

File: examples/fabric/reinforcement_learning/train_torch.py (hunk 1 above)
Comment (ryan597) [reply examples/fabric/reinforcement_learning/train_torch.py]: Thanks!  Didn't know about the tensor/Tensor convention.  
Made two commits, first for the suggestions you gave and then there's one or two others in places I hadn't previously written, so we can remove the `from torch import Tensor` line too.

File: examples/fabric/reinforcement_learning/rl/utils.py [hunk 6]
```diff
@@ -119,7 +117,12 @@ def parse_args():
     return args
 
 
-def layer_init(layer: torch.nn.Module, std: float = np.sqrt(2), bias_const: float = 0.0, ortho_init: bool = True):
+def layer_init(
+    layer: torch.nn.Module,
+    std: float = torch.sqrt(torch.tensor(2)).item(),
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/rl/utils.py:122]: how about `math.sqrt(2)`?

File: examples/fabric/reinforcement_learning/rl/agent.py [hunk 7]
```diff
@@ -24,7 +23,8 @@ def __init__(self, envs: gym.vector.SyncVectorEnv, act_fun: str = "relu", ortho_
             raise ValueError("Unrecognized activation function: `act_fun` must be either `relu` or `tanh`")
         self.critic = torch.nn.Sequential(
             layer_init(
-                torch.nn.Linear(np.array(envs.single_observation_space.shape).prod(), 64), ortho_init=ortho_init
+                torch.nn.Linear(int(torch.tensor(envs.single_observation_space.shape).prod()), 64),
```
Comment (awaelchli) [inline examples/fabric/reinforcement_learning/rl/agent.py:26]: math.prod() will be a tiny bit shorter and avoids conversions :)

**Review by awaelchli — APPROVED** [review]

File: examples/fabric/reinforcement_learning/rl/utils.py (hunk 6 above)
Comment (ryan597) [reply examples/fabric/reinforcement_learning/rl/utils.py]: Changed, thanks. (I was being silly and didn't want to introduce the math import when I was trying to remove the numpy import)

Comment (awaelchli) [reply examples/fabric/reinforcement_learning/rl/utils.py]: Since math is a built-in module of Python, this is not a problem at all :)

**Comment by awaelchli** [comment]
> I'm not sure how to go about resolving the failed test as the examples I changed aren't going into that test..

They are unrelated, you can ignore them and we will rerun them before merging.

**Review by justusschock — APPROVED** [review]
